protocol.md

```diff
@@ -0,0 +1,752 @@
+`0xc0` | cancel            | Cancel an outstanding call req / forward req (no body)
+`0xc1` | claim             | Claim / cancel a redundant request
+`0xd0` | ping req          | Protocol level ping req (no body)
+`0xd1` | ping res          | Ping res (no body)
+`0xff` | error             | Protocol level error.
+
+The framing layer is common to all payloads. It intentionally limits the frame
+size to 64KiB to allow better interleaving of frames across a shared TCP
+connection. In order to handle most operations implementations will need to
+decode some part of the payload.
+
+### id:4
+
+An identifier for this message that is chosen by the sender of a request. This
+id is only valid for this sender on this connection. This is similar to how TCP
+has a sequence number in each direction. Each side of a connection may happen
+to select overlapping message ids, which is fine because they are directional.
+
+`id` represents the top level message id. Both request frames and response
+frames use the same id, which is how they are matched up. A single id may be
+used across multiple request or response frames after being broken up into a
+sequence of fragments, as described below.
+
+Valid values for "id" are from `0` to `0xFFFFFFFE`. The value `0xFFFFFFFF` is
+reserved for protocol error responses.
+
+### reserved:8
+
+Unused space for future protocol revisions, padded for read alignment, which
+may or may not help on modern Intel processors.
+
+### payload
+
+0 or more bytes whose contents are determined by the frame type.
+
+The leng of the payload is frame `size` - 16.
+
+Full details on each payload body follow below.
+
+## Payload: init req (type 0x01)
+
+Schema: `version:2 nh:2 (key~2 value~2){nh}`
+
+This must be the first message sent on a new connection. It is used to
+negotiate a common protocol version and describe the service names on both
+ends. In the future, we will likely use this to negotiate authentication and
+authorization between services.
+
+### version
+
+`version` is a 16 bit number. The currently specified protocol version is 2. If
+new versions are required, this is where a common version can be negotiated.
+
+### headers
+
+There are a variable number of key/value pairs. For version 2, the following
+are required:
+
+`host\_port`    | address:port where this process can be reached
+`process\_name` | additional identifier for this instance, used for logging
+
+For forward compatibility, implementations should ignore headers other than
+these.
+
+## Payload: init res (type 0x02)
+
+Schema: `version:2 nh:2 (key~2 value~2){nh}`
+
+The initiator requests a version number, and the server responds with the
+actual version that will be used for the rest of this connection. The
+name/values are the same, but identify the server.
+
+## Payload: call req (type 0x03)
+
+Schema: `flags:1 ttl:4 tracing:24 traceflags:1 service~2 nh:1 (hk~1 hv~1){nh} csumtype:1 (csum:4){0,1} arg1~2 arg2~2 arg3~2`
```

Comment:
This is too long. This can probably be changed into a code block split with the schema split across multiple lines.